File: 75Review/BinarySearch/74.cpp
#include <vector>
using namespace std;

class Solution {
  public:
    bool searchMatrix(vector<vector<int>> &matrix, int target) {
        int n = matrix.size(), m = matrix[0].size();
        int l = 0, r = (n * m) - 1;

        while (l <= r) {
            int mid = l + (r - l / 2); // 1 + 5
            int x = mid / m;
            int y = mid % m;

            if (matrix[x][y] == target) return true;
            if (matrix[x][y] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }

        return false;
    }
};

File: src/2025/rusting/74.rs
struct Solution;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let rows = matrix.len();
        let cols = matrix[0].len();
        let mut l: i32 = 0;
        let mut r: i32 = rows as i32 * cols as i32 - 1;

        while l <= r {
            let m = (l + (r  - l) / 2) as usize;
            let (m_x, m_y) = (m / cols, m % cols);

            if matrix[m_x][m_y] == target {
                return true;
            } else if target < matrix[m_x][m_y] {
                r = m as i32 - 1;
            } else {
                l = m as i32 + 1;
            }
        }
        false
    }
}